databricks/koalas/frame.py

```diff
@@ -1063,6 +1063,58 @@ def notnull(self):
 
     notna = notnull
 
+    def nunique(self, axis: int = 0, dropna: bool = True, approx: bool = False,
+                rsd: float = 0.05) -> pd.Series:
+        """
+        Return number of unique elements in the object.
+
+        Excludes NA values by default.
+
+        Parameters
+        ----------
+        axis : int, default 0
+            Can only be set to 0 at the moment.
+        dropna : bool, default True
+            Don’t include NaN in the count.
+        approx: bool, default False
+            If False, will use the exact algorithm and return the exact number of unique.
+            If True, it uses Spark's approximate algorithm, which is faster in most circumstances.
+            Note: this parameter is specific to Spark and is not found in pandas.
+        rsd: float, default 0.05
+            Maximum estimation error allowed. Just like ``approx`` this parameter is specific to
+            Spark.
+
+        Returns
+        -------
+        The number of unique values per column as a pandas Series.
+
+        Examples
+        --------
+        >>> ks.DataFrame({'A': [1, 2, 3], 'B': [np.nan, 3, np.nan]}).nunique()
+        A    3
+        B    1
+        Name: 0, dtype: int64
+
+        >>> ks.DataFrame({'A': [1, 2, 3], 'B': [np.nan, 3, np.nan]}).nunique(dropna=False)
+        A    3
+        B    2
+        Name: 0, dtype: int64
+        """
+        if axis != 0:
+            raise ValueError("The 'nunique' method only works with axis=0 at the moment")
+        count_fn = partial(F.approx_count_distinct, rsd=rsd) if approx else F.countDistinct
+        if dropna:
+            res = self._sdf.select([count_fn(Column(c))
+                                   .alias(c)
+                                    for c in self.columns])
+        else:
+            res = self._sdf.select([(count_fn(Column(c))
+                                     + F.when(F.count(F.when(F.col(c).isNull(), 1).otherwise(None))
```

Comment:
i find the when/when/otherwise pretty difficult to understand. can you add some comment explaining them above? maybe add some pseudocode ...

Reply:
Fair enough. 295c6c4 adds some comments that explain the rationale behind writing the select statement this way.